specs/core/1_beacon-chain.md

```diff
@@ -0,0 +1,686 @@
+    for i, custody_bits in enumerate(indexed_attestation.attestation.custody_bits):
+        assert len(custody_bits) == len(indexed_attestation.committee)
+        for participant, abit, cbit in zip(indexed_attestation.committee, aggregation_bits, custody_bits):
+            if abit:
+                all_pubkeys.append(state.validators[participant].pubkey)
+                # Note: only 2N distinct message hashes
+                all_message_hashes.append(hash_tree_root(
+                    AttestationCustodyBitWrapper(hash_tree_root(indexed_attestation.data), i, cbit)
+                ))
+            else:
+                assert cbit == False
+        
+    return bls_verify_multiple(
+        pubkeys=all_pubkeys,
+        message_hashes=all_message_hashes,
+        signature=indexed_attestation.signature,
+        domain=get_domain(state, DOMAIN_BEACON_ATTESTER, indexed_attestation.data.target.epoch),
+    )
+```
+
+
+### Block processing
+
+```python
+def process_block(state: BeaconState, block: BeaconBlock) -> None:
+    process_block_header(state, block)
+    process_randao(state, block.body)
+    process_eth1_data(state, block.body)
+    verify_shard_transition_false_positives(state, block)
+    process_light_client_signatures(state, block)
+    process_operations(state, block.body)
+```
+
+
+#### Operations
+
+```python
+def process_operations(state: BeaconState, body: BeaconBlockBody) -> None:
+    # Verify that outstanding deposits are processed up to the maximum number of deposits
+    assert len(body.deposits) == min(MAX_DEPOSITS, state.eth1_data.deposit_count - state.eth1_deposit_index)
+    
+    def for_ops(operations, fn):
+        for operation in operations:
+            fn(state, operation)
+    
+    for_ops(body.proposer_slashings, process_proposer_slashing)
+    for_ops(body.attester_slashings, process_attester_slashing)
+
+    # New attestation processing
+    process_attestations(state, body, body.attestations)
+
+    for_ops(body.deposits, process_deposit)
+    for_ops(body.voluntary_exits, process_voluntary_exit)
+
+    # See custody game spec.
+    process_custody_game_operations(state, body)
+
+    # TODO process_operations(body.shard_receipt_proofs, process_shard_receipt_proofs)
+```
+
+##### New Attestation processing
+
+###### `validate_attestation`
+
+```python
+def validate_attestation(state: BeaconState, attestation: Attestation) -> None:
+    data = attestation.data
+    assert data.index < ACTIVE_SHARDS
+    shard = get_shard(state, attestation)
+    proposer_index = get_beacon_proposer_index(state)
+
+    # Signature check
+    assert is_valid_indexed_attestation(state, get_indexed_attestation(state, attestation))
+    # Type 1: on-time attestations
+    if attestation.custody_bits != []:
+        # Correct slot
+        assert data.slot == state.slot
+        # Correct data root count
+        assert len(attestation.custody_bits) == len(get_offset_slots(state, state.shard_next_slots[shard]))
```

Comment:
`state.shard_next_slots` is not defined in the BeaconState definition above

Reply:
Removed `shard_next_slots` and added `get_next_slot_for_shard`. The value can be derived from `state.shard_transitions[shard].slot`.